### ISSUE TITLE
Mark callstack relevant methods with {-# NOINLINE log... #-} - otherwise ghc-option fspecialise-aggressively leads to wrong callstack information

### ISSUE THREAD

**Issue by alaendle** [kowainik/co-log#209]
E.g. 

```haskell
example :: WithLog env Message m => m ()
example = do
    Colog.log Debug "debug"
    Colog.log Info "info"
    log2 Warning "adsfdas"
    log2 Error "adsfdas"

log2 :: WithLog env (Msg sev) m => sev -> Text -> m ()
log2 msgSeverity msgText =
    withFrozenCallStack (logMsg Msg{ msgStack = callStack, .. })
{-# NOINLINE log2 #-}
```
leads to
```
[Debug]   [21 Oct 2020 16:12:55.974 +00:00] [Main.example#112] [ThreadId 3] debug
[Info]    [21 Oct 2020 16:12:55.975 +00:00] [Main.example#112] [ThreadId 3] info
[Warning] [21 Oct 2020 16:12:55.975 +00:00] [Main.example#113] [ThreadId 3] adsfdas
[Error]   [21 Oct 2020 16:12:55.976 +00:00] [Main.example#114] [ThreadId 3] adsfdas
```

Please note the identical line number for the first two log entries.
I'm not really sure if this is an issue of co-log or GHC - nonetheless it could be avoided by adding a NOINLINE pragma as `log2` demonstrates.

**Comment by chshersh** [comment]
@alaendle That's an interesting observation :thinking: I would start with opening an issue in GHC tracker to start a discussion about this behaviour. When optimization changes the behaviour, it's usually a bug, and I know from experience that such bugs sometimes happen with GHC. So, it would be good to know, whether it's an actual bug or expected behaviour. After that we can act accordingly :slightly_smiling_face:

**Comment by alaendle** [comment]
After taking a even closer look this appears to be inconsistent in GHC, so as @chshersh recommended I've created a corresponding GHC issue: https://gitlab.haskell.org/ghc/ghc/-/issues/18751
I'm exited if this considered as a bug or expected behaviour, nonetheless one could consider to add the NOINLINE-pragmas as an interim solution. But lets wait for an answer of the GHC folk first.

**Comment by alaendle** [comment]
The problem is solved with ghc-9.0.1 - so I think this issue can be closed since the problem will disappear with the adaption of newer compilers.

**Comment by chshersh** [comment]
@alaendle Thanks for investigating this issue! ❤️
